File: content/examples/other/leetcode/problems/valid-parentheses/example-2/solution.js
function isValidParenthesesFast(s) {
  const stack = [];
  const closingToOpening = new Map([
    [")", "("],
    ["]", "["],
    ["}", "{"],
  ]);

  for (let i = 0; i < s.length; i += 1) {
    const ch = s[i];
    if (closingToOpening.has(ch)) {
      const open = stack.pop();
      if (open !== closingToOpening.get(ch)) return false;
    } else {
      stack.push(ch);
    }
  }
  return stack.length === 0;
}

if (require.main === module) {
  console.log(isValidParenthesesFast("()[]{}"));
  console.log(isValidParenthesesFast("([)]"));
}

module.exports = { isValidParenthesesFast };

File: content/examples/other/leetcode/problems/valid-parentheses/example-2/solution.py
def is_valid_parentheses_fast(s):
    stack = []
    closing_to_opening = {")": "(", "]": "[", "}": "{"}
    for ch in s:
        if ch in closing_to_opening:
            if not stack or stack.pop() != closing_to_opening[ch]:
                return False
        else:
            stack.append(ch)
    return len(stack) == 0


if __name__ == "__main__":
    print(is_valid_parentheses_fast("()[]{}"))
    print(is_valid_parentheses_fast("([)]"))
